File: LeetCode/148.sort-list.cpp
/*
 * @lc app=leetcode id=148 lang=cpp
 *
 * [148] Sort List
 */
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class SolutionA
{
    // Merge sort elements between beg and end (excluding end)
    // Not elegant enough
    ListNode *mergeSort(ListNode *beg, ListNode *end)
    {
        if (beg == end)
            return nullptr;

        if (beg->next == end)
            return beg;

        // Find the middle point
        ListNode *slow = beg, *fast = beg;
        while (fast != end && fast->next != end)
        {
            slow = slow->next;
            fast = fast->next->next;
        }

        // Merge sort the left part
        ListNode *left = mergeSort(beg, slow);

        // Merge sort the right part
        ListNode *right = mergeSort(slow, end);

        // Merge both parts
        ListNode dummy(0);
        ListNode *prev = &dummy;
        while (left != slow && right != end)
        {
            if (left->val <= right->val)
            {
                prev->next = left;
                left = left->next;
            }
            else
            {
                prev->next = right;
                right = right->next;
            }
            prev = prev->next;
        }
        while (left != slow)
        {

            prev->next = left;
            left = left->next;
            prev = prev->next;
        }

        while (right != end)
        {

            prev->next = right;
            right = right->next;
            prev = prev->next;
        }
        prev->next = end;
        return dummy.next;
    }

public:
    ListNode *sortList(ListNode *head)
    {
        return mergeSort(head, nullptr);
    }
};

class Solution
{
public:
    ListNode *sortList(ListNode *head)
    {
        if (head == nullptr || head->next == nullptr)
            return head;

        // Find the middle point
        ListNode *slow = head, *fast = head->next;
        while (fast && fast->next)
        {
            slow = slow->next;
            fast = fast->next->next;
        }

        // Sort the two parts accordingly
        
        ListNode *right = sortList(slow->next);
        // Cut the list off from the middle element
        slow->next = nullptr;

        ListNode *left = sortList(head);

        // Merge two parts.
        ListNode dummy(0);
        ListNode *prev = &dummy;
        while (left && right)
        {
            if (left->val <= right->val)
            {
                prev->next = left;
                left = left->next;
            }
            else
            {
                prev->next = right;
                right = right->next;
            }
            prev = prev->next;
        }
        left = left ? left : right;
        prev->next = left;
        return dummy.next;
    }
};

File: LeetCode/java/148.sort-list.java
/*
 * @lc app=leetcode id=148 lang=java
 *
 * [148] Sort List
 *
 * https://leetcode.com/problems/sort-list/description/
 *
 * algorithms
 * Medium (41.69%)
 * Likes:    3052
 * Dislikes: 140
 * Total Accepted:    281.2K
 * Total Submissions: 654.3K
 * Testcase Example:  '[4,2,1,3]'
 *
 * Sort a linked list in O(n log n) time using constant space complexity.
 * 
 * Example 1:
 * 
 * 
 * Input: 4->2->1->3
 * Output: 1->2->3->4
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: -1->5->3->4->0
 * Output: -1->0->3->4->5
 * 
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode sortList(ListNode head) {
        return mergeSortList(head);
    }
    private ListNode mergeSortList(ListNode head){
        if(head == null||head.next == null)
            return head;
        ListNode dummy = new ListNode(0,head);
        ListNode slow = dummy, fast = head;
        while(fast!=null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }
        fast = slow.next;
        slow.next = null;
        ListNode left = mergeSortList(head);
        ListNode right = mergeSortList(fast);
        ListNode merged = mergeList(left,right);
        return merged;
    }

    private ListNode mergeList(ListNode left, ListNode right){
        if(left == null)
            return right;
        if(right == null)
            return left;
        if(left.val<=right.val){
            left.next=mergeList(left.next, right);
            return left;
        }else{
            right.next=mergeList(left, right.next);
            return right;
        }
    }
}
// @lc code=end

